node/core/approval-voting/src/lib.rs

```diff
@@ -774,6 +776,145 @@ where
+	overlay_db: &mut OverlayedBackend<'_, impl Backend>,
+	state: &mut State,
+) -> SubsystemResult<Vec<Action>> {
+	let all_candidates = overlay_db.load_all_candidates().map_err(|e| {
+		tracing::error!(target: LOG_TARGET, "Failed initial load of active candidates: {:?}", e);
+		e
+	})?;
+
+	let mut actions = Vec::new();
+	for candidate in all_candidates.into_iter() {
+		let mut assigned_but_unapproved_candidates = Vec::new();
+		let session = candidate.session;
+		let info = match state.session_info(session) {
+			Some(info) => info,
+			None => {
+				tracing::debug!(target: LOG_TARGET, "Missing session info for {:?}", session);
+				continue
+			},
+		};
+
+		// Compute the local validator indexes, if any, for this session.
+		let local_validators: Vec<_> = info
+			.validators
+			.iter()
+			.enumerate()
+			.filter(|(_, validator)| {
+				state
+					.keystore
+					.key_pair::<ValidatorPair>(validator)
+					.ok()
+					.map_or(false, |v| v.is_some())
+			})
+			.map(|(index, _)| ValidatorIndex(index as _))
+			.collect();
+
+		// Compute the local assignment indexes, if any, for this session.
+		let local_assigners: Vec<_> = info
+			.assignment_keys
+			.iter()
+			.enumerate()
+			.filter(|(_, assigner)| {
+				state
+					.keystore
+					.key_pair::<AssignmentPair>(assigner)
+					.ok()
+					.map_or(false, |v| v.is_some())
+			})
+			.map(|(index, _)| ValidatorIndex(index as _))
+			.collect();
+
+		let is_session_validator = !local_validators.is_empty();
+		let is_session_assigner = !local_assigners.is_empty();
+
+		// Determine whether or not the local node controls session validation and assignment. If
+		// not, skip all candidates in this session.
+		let controls_validation_and_assignment = is_session_validator && is_session_assigner;
+		if !controls_validation_and_assignment {
+			continue
+		}
+
+		// Determine the candidates in this session that need to have their approvals resumed.
+		if !local_validators.iter().all(|index| candidate.has_approved(*index)) {
+			let candidate_receipt = candidate.candidate.clone();
+			assigned_but_unapproved_candidates.extend(
+				candidate
+					.block_assignments
+					.into_iter()
+					// Retain any candidates that have our assignment indexes in the assignments
+					// bitfield.
+					.filter(|(_, approval_entry)| {
+						local_assigners.iter().any(|index| approval_entry.is_assigned(*index))
+					})
+					.map(|(block_hash, approval_entry)| {
+						(candidate_receipt.clone(), block_hash, approval_entry)
+					})
+					.collect::<Vec<_>>(),
+			);
+		}
+
+		// Construct the LaunchApproval action for each candidate.
+		for (candidate_receipt, block_hash, approval_entry) in
+			assigned_but_unapproved_candidates.into_iter()
+		{
+			// Skip any candidates for which we don't have our assignment triggered.
+			let (cert, assignment_tranche, validator) = match approval_entry.our_assignment() {
+				Some(our) => (our.cert().clone(), our.tranche(), our.validator_index()),
```

Comment:
nit: not really a need to complicate this, the tuple extraction can be done afterwards